src/io/stm_tests.ml

```diff
@@ -0,0 +1,298 @@
+    { path; channel }
+
+  let cleanup { path; channel } =
+    if channel <> Stdlib.stdout
+    then Out_channel.close channel;
+    (*Printf.printf "Removing %s\n%!" path;*)
+    Sys.remove path
+
+  let precond c s = match c,s with
+    | Open_text, Closed _ -> true
+    | Open_text, Open _ -> false
+    | _, Open _ -> true
+    | _, _ -> false
+
+  let run c ({path;channel = oc} as r) = match c with
+    | Open_text       -> Res (result unit exn, protect (fun path -> (r.channel <- Out_channel.open_text path;())) path)
+    | Seek p          -> Res (unit, Out_channel.seek oc p)
+    | Pos             -> Res (result int64 exn, protect Out_channel.pos oc)
+    | Length          -> Res (int64, Out_channel.length oc)
+    | Close           -> Res (result unit exn, protect Out_channel.close oc)
+    | Close_noerr     -> Res (unit, Out_channel.close_noerr oc)
+    | Flush           -> Res (unit, Out_channel.flush oc)
+    | Output_char c   -> Res (result unit exn, protect (Out_channel.output_char oc) c)
+    | Output_byte i   -> Res (result unit exn, protect (Out_channel.output_byte oc) i)
+    | Output_string s -> Res (result unit exn, protect (Out_channel.output_string oc) s)
+    | Output_bytes b  -> Res (result unit exn, protect (Out_channel.output_bytes oc) b)
+    | Output (b,p,l)  -> Res (result unit exn, protect (Out_channel.output oc b p) l)
+    | Output_substring (s,p,l) -> Res (result unit exn, protect (Out_channel.output_substring oc s p) l)
+    | Set_binary_mode b ->
+      if Sys.win32 || Sys.cygwin
+      then Res (unit, (Out_channel.flush oc; Out_channel.set_binary_mode oc b)) (* flush before changing mode *)
+      else Res (unit, Out_channel.set_binary_mode oc b)
+    | Set_buffered b  -> Res (unit, Out_channel.set_buffered oc b)
+    | Is_buffered     -> Res (bool, Out_channel.is_buffered oc)
+
+  let postcond c (s:state) res = match c, res with
+    | Open_text, Res ((Result (Unit,Exn),_), r) ->
+       (match s,r with
+        | Closed _, Ok ()
+        | Closed _, Error (Sys_error _) (*"/tmp/lin-dsl-03ba23: Too many open files"*)
+        | Open _, Ok ()
+        | Open _, Error (Sys_error _) -> true
+        | _ -> false)
+    | Seek _, Res ((Unit,_), ()) -> true
+    | Pos, Res ((Result (Int64,Exn),_), r) ->
+       (match s with
+        | Closed _ -> true (*r = Error (Invalid_argument "Pos exception") - unspecified *)
+        | Open { position; length = _; buffered = _; binary_mode = _ } -> r = Ok position)
+    | Length, Res ((Int64,_),i) ->
+       (match s with
+        | Closed _ -> true
+        | Open { position = _; length; buffered = _; binary_mode = _ } -> i <= length)
+    | Close, Res ((Result (Unit,Exn),_), r) ->
+       (match s,r with
+         | Closed _, Error (Sys_error _) (*"Close exception" - unspecified *)
+         | Open _, Ok () -> true
+         | _ -> false)
+    | Close_noerr, Res ((Unit,_), r) ->
+       (match s,r with
+         | Closed _, ()
+         | Open _, () -> true)
```

Comment:
This might be squashed to `true`.

Reply:
This one I needed to change and extend in 8236876